contrib/babelfishpg_common/src/datetime.c

```
@@ -517,6 +518,116 @@ tsql_decode_datetime_fields(char *orig_str, char *str, char **field, int nf, int
 	return 0;
 }
 
+/*
+ * Apply datetime rounding
+ * This implements the same logic as TdsTimeDifferenceDatetime
+ */
+
+static void
+handle_datetime_carry_over(struct pg_tm *tm, int *rounded_msec)
+{
+	int leap, max_days;
+	
+	if (*rounded_msec >= 1000)
+	{
+		*rounded_msec -= 1000;
+		tm->tm_sec++;
+		
+		/* Handle cascading overflows */
+		if (tm->tm_sec >= 60)
+		{
+			tm->tm_sec = 0;
+			tm->tm_min++;
+			
+			if (tm->tm_min >= 60)
+			{
+				tm->tm_min = 0;
+				tm->tm_hour++;
+				
+				if (tm->tm_hour >= 24)
+				{
+					tm->tm_hour = 0;
+					tm->tm_mday++;
+					
+					/* Handle month/year rollover */
+					leap = isleap(tm->tm_year) ? 1 : 0;
+					max_days = day_tab[leap][tm->tm_mon - 1];
+					
+					if (tm->tm_mday > max_days)
+					{
+						tm->tm_mday = 1;
+						tm->tm_mon++;
+						if (tm->tm_mon > 12)
+						{
+							tm->tm_mon = 1;
+							tm->tm_year++;
+						}
+					}
+				}
+			}
+		}
+	}
+}
+
+Timestamp
+roundoff_datetime(Timestamp timestamp)
+{
+	struct pg_tm tm;
+	fsec_t		fsec;
+	int			msec, rounded_msec = 0;
+	Timestamp	result;
+	int			unit, base;
+
+	if (TIMESTAMP_NOT_FINITE(timestamp))
+		return timestamp;
```

Should we throw error such as "Datetime out of range" instead of returning timestamp as it is ?

---

+1

---

Updated in latest rev